psi4/src/psi4/libmints/mintshelper.cc

```diff
@@ -2597,6 +2597,230 @@ std::vector<SharedMatrix> MintsHelper::ao_overlap_kinetic_deriv2_helper(const st
     return grad;
 }
 
+/* 1st derivatives of electric dipole integrals in the AO basis */
+std::vector<SharedMatrix> MintsHelper::ao_elec_dip_deriv1_helper(int atom) {
+    std::vector<std::string> cartcomp;
+    cartcomp.push_back("X");
+    cartcomp.push_back("Y");
+    cartcomp.push_back("Z");
+
+    std::shared_ptr<OneBodyAOInt> Dint(integral_->ao_dipole(1));
+
+    std::shared_ptr<BasisSet> bs1 = Dint->basis1();
+    std::shared_ptr<BasisSet> bs2 = Dint->basis2();
+
+    int nbf1 = bs1->nbf();
+    int nbf2 = bs2->nbf();
+
+    std::vector<SharedMatrix> grad;
+    for (int p = 0; p < 3; p++) {
+        std::stringstream sstream;
+        sstream << "ao_mu" << cartcomp[p] << "_deriv1_";
+        for (int q = 0; q < 3; q++) {
+            sstream << atom << cartcomp[q];
+            grad.push_back(SharedMatrix(new Matrix(sstream.str(), nbf1, nbf2)));
+        }
+    }
+
+    const double *buffer = Dint->buffer();
+
+    for (int P = 0; P < bs1->nshell(); P++) {
+        for (int Q = 0; Q < bs2->nshell(); Q++) {
+            int nP = basisset_->shell(P).nfunction();
+            int oP = basisset_->shell(P).function_index();
+            int aP = basisset_->shell(P).ncenter();
+
+            int nQ = basisset_->shell(Q).nfunction();
+            int oQ = basisset_->shell(Q).function_index();
+            int aQ = basisset_->shell(Q).ncenter();
+
+            if (aP != atom && aQ != atom) continue;
+
+            Dint->compute_shell_deriv1(P, Q);
+
+//            for (int mu_cart = 0; mu_cart < 3; mu_cart++) {
+//                for (int atom_cart = 0; atom_cart < 3; atom_cart++) {
+//                    int offset = (6 * mu_cart * nP * nQ) + (atom_cart * nP * nQ) + (3 * (atom == aQ) * nP * nQ);
+//                    for (int p = 0; p < nP; p++) {
+//                        for (int q = 0; q < nQ; q++) {
+//                            grad[3 * mu_cart + atom_cart]->add(p + oP, q + oQ, buffer[p * nQ + q + offset]);
+//                            if (aP == aQ) grad[3 * mu_cart + atom_cart]->add(p + oP, q + oQ, buffer[p * nQ + q + offset - (3 * nP * nQ)]);
+//                        }
+//                    }
+//                }
+//            }
```

Comment:
@dgasmith @loriab Quick question: I wrote the code in two equivalent forms - 1. the above commented code that exhibits a more compact loop structure, and 2. the code below this that mirrors a similar structure to other surrounding mintshelper functions. Is one of these preferred over the other? I tend to lean toward fewer lines of code being better, but perhaps it's better to match the style of existing similar functions. Thoughts?

Reply:
@jturney @andysim Probably have more preferences here.

Reply:
I'm good with the compact code.